consensus/src/quorum_store/batch_generator.rs

```diff
@@ -263,33 +274,42 @@ impl BatchGenerator {
-                                self.config.back_pressure.dynamic_min_batch_count,
+                            back_pressure_decrease_latest = now;
+                            dynamic_pull_txn_per_s = std::cmp::max(
+                                (dynamic_pull_txn_per_s as f64 * self.config.back_pressure.decrease_fraction) as u64,
+                                self.config.back_pressure.dynamic_min_txn_per_s,
                             );
-                            debug!("QS: dynamic_max_pull_count: {}", dynamic_max_pull_count);
+                            debug!("QS: dynamic_max_pull_txn_per_s: {}", dynamic_pull_txn_per_s);
                         }
                         counters::QS_BACKPRESSURE.set(1);
-                        counters::QS_BACKPRESSURE_DYNAMIC_MAX.set(dynamic_max_pull_count as i64);
                     } else {
                         // additive increase, every second
                         if back_pressure_increase_latest.elapsed() >= back_pressure_increase_duration {
-                            back_pressure_increase_latest = Instant::now();
-                            dynamic_max_pull_count = std::cmp::min(
-                                dynamic_max_pull_count + self.config.back_pressure.dynamic_min_batch_count,
-                                self.config.back_pressure.dynamic_max_batch_count,
+                            back_pressure_increase_latest = now;
+                            dynamic_pull_txn_per_s = std::cmp::min(
+                                dynamic_pull_txn_per_s + self.config.back_pressure.dynamic_min_txn_per_s,
+                                self.config.back_pressure.dynamic_max_txn_per_s,
                             );
-                            debug!("QS: dynamic_max_pull_count: {}", dynamic_max_pull_count);
+                            debug!("QS: dynamic_max_pull_txn_per_s: {}", dynamic_pull_txn_per_s);
                         }
                         counters::QS_BACKPRESSURE.set(0);
-                        counters::QS_BACKPRESSURE_DYNAMIC_MAX.set(dynamic_max_pull_count as i64);
                     }
-                    if let Some(proof_rx) = self.handle_scheduled_pull(dynamic_max_pull_count).await {
-                        proofs_in_progress.push(Box::pin(proof_rx));
+                    let since_last_pull_ms = std::cmp::min(
+                        now.duration_since(last_non_empty_pull).as_millis(),
+                        self.config.batch_generation_max_interval_ms as u128
+                    ) as usize;
+                    if !self.back_pressure.batch_count || since_last_pull_ms == self.config.batch_generation_max_interval_ms {
+                        last_non_empty_pull = now;
+                        let dynamic_pull_max_txn = std::cmp::max(
+                            (since_last_pull_ms as f64 / 1000.0 * dynamic_pull_txn_per_s as f64) as u64, 1);
+                        counters::QS_BACKPRESSURE_DYNAMIC_MAX.set(dynamic_pull_max_txn as i64);
```

Comment:
probably worth to split the dynamic_pull_txn_per_s and the since_last_pull_ms?

Reply:
also, thinking out loud, is it better to log here, or only when there actually is a batch that we create? (i.e. move it inside the if below?). If inside we will get average stats based on when the batch is created

also maybe having a separate boolean counter for  counter self.back_pressure.txn_count and self.back_pressure.batch_count, so we can see which back pressure is triggered?

Reply:
I think it makes sense to understand how the value is changed, regardless of whether there were txns to pull or not.

I updated to have separate boolean counters